File: apps/mybookkeeper/frontend/src/shared/types/utility/utility-plan-rate-comparison-row.ts
import type { UtilityPlanBenchmarkStatus } from "@/shared/types/utility/utility-plan-benchmark-status";
import type { UtilityPlanSummary } from "@/shared/types/utility/utility-plan-summary";

/**
 * One plan measured against the market benchmark for its service type.
 *
 * Mirrors ``schemas/properties/utility_plan_rate_comparison_row.py``.
 * ``plan_figure`` and ``benchmark_figure`` share a unit that depends on the
 * plan's service type: ¢/kWh for metered service, cents per month for flat.
 */
export interface UtilityPlanRateComparisonRow {
  plan: UtilityPlanSummary;
  status: UtilityPlanBenchmarkStatus;
  plan_figure: string | null;
  benchmark_figure: string | null;
  /** Signed — negative means the plan beats the market. */
  gap_pct: string | null;
  benchmark_is_stale: boolean;
}

File: apps/mybookkeeper/backend/app/schemas/properties/utility_plan_rate_comparison_row.py
"""One plan measured against the market benchmark for its service type."""
from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.schemas.properties.utility_plan_summary import UtilityPlanSummary


class UtilityPlanRateComparisonRow(BaseModel):
    plan: UtilityPlanSummary
    # One of app.core.market_benchmark_constants.BENCHMARK_STATUSES.
    status: str
    # Both figures are in the unit the service type is compared in — ¢/kWh for
    # metered service, cents per month for flat service. The client reads the
    # unit from the plan's service_type rather than being told twice.
    plan_figure: Decimal | None = None
    benchmark_figure: Decimal | None = None
    # Signed: negative means the plan beats the market, which is worth showing
    # as the evidence that an earlier switch worked.
    gap_pct: Decimal | None = None
    # True when the benchmark behind this row is old enough that the comparison
    # deserves a caveat rather than a confident badge.
    benchmark_is_stale: bool = False

    model_config = ConfigDict(from_attributes=True)
